### PR TITLE
add capability to analyze aggregates to reports

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #xxxx .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.
  - [x] update pr name if this turns into the right approach

Working on getting aggregates into reports. I think the proposed code mostly does the job, but I can't say I like any of it. Maybe good enough for now? Open questions:

- [x] `template.md` table needs to be updated to account for `/aggregates` instead of `/observations`. How should that info get to the template?
- [x] does the report need to change in any other way for an aggregate? I've not yet successfully rendered one, but I don't think so
- [x] should we make an `APISession.get_values(datamodel_obj, start, end)` method to replace the `if/else` block in `main.get_data_for_report`? Main issue with that kind of function is that forecast data has a different structure than observation and aggregate data. Another option could be adding `get_values(session, start, end)` methods to the datamodel objects, but that's inconsistent with our current approach.
- [x] can we add a live aggregate or two that the test user can access? 
- [x] are there issues with quality flags that should be addressed now?

what do you think @alorenzo175?

### CODE REVIEW

**Comment by wholmgren** [comment]
Specifying `object_pairs` in report metadata is also a problem. 

```
        "object_pairs": [
            ["39220780-76ae-4b11-bef1-7a75bdc784e3", "458ffc27-df0b-11e9-b622-62adb5fd6af0"],  # 458... is an aggregate, but APISession.process_report_dict tries to hit observations/458...
            ["11c20780-76ae-4b11-bef1-7a75bdc784e3", "123e4567-e89b-12d3-a456-426655440000"]
      ]
```

**Comment by wholmgren** [comment]
what about

```
        "object_pairs": [
            ["forecast": "39220780-76ae-4b11-bef1-7a75bdc784e3", 
             "aggregate": "458ffc27-df0b-11e9-b622-62adb5fd6af0"],  
            ["forecast": "11c20780-76ae-4b11-bef1-7a75bdc784e3", 
             "observation": "123e4567-e89b-12d3-a456-426655440000"]
      ]

File: solarforecastarbiter/reports/main.py [hunk 1]
```diff
@@ -85,13 +85,21 @@ def get_data_for_report(session, report):
...
         # only get the raw data once.
         forecast_id = fxobs.forecast.forecast_id
-        observation_id = fxobs.observation.observation_id
         if fxobs.forecast not in data:
             data[fxobs.forecast] = session.get_forecast_values(
                 forecast_id, report.start, report.end)
         if fxobs.observation not in data:
-            data[fxobs.observation] = session.get_observation_values(
-                observation_id, report.start, report.end)
+            # alternative: add APISession.get_values(datamodel_obj, start, end)
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/main.py]: yes

File: solarforecastarbiter/datamodel.py [hunk 2]
```diff
@@ -870,7 +870,7 @@ class ForecastObservation(BaseModel):
     __post_init__ checking.
     """
     forecast: Forecast
-    observation: Observation
+    observation: Union[Observation, Aggregate]
```
Comment (alorenzo175) [inline solarforecastarbiter/datamodel.py]: observation can be observation or aggregate? what about a ``ForecastAggregate`` object and then report can take either of those. Then processing the ``object_pairs``like you suggested probably works better

Comment (wholmgren) [reply solarforecastarbiter/datamodel.py]: Was just wondering about that too. Could we keep the attribute  `observation` or would you want to change it to `aggregate`? Lots of code reaches for that attribute.

File: solarforecastarbiter/reports/main.py (hunk 1 above)
Comment (alorenzo175) [reply solarforecastarbiter/reports/main.py]: actually not so sure

**Comment by alorenzo175** [comment]
> template.md table needs to be updated to account for /aggregates instead of /observations. How should that info get to the template?

I guess putting in a tuple isn't bad

> does the report need to change in any other way for an aggregate? I've not yet successfully rendered one, but I don't think so

> should we make an APISession.get_values(datamodel_obj, start, end) method to replace the if/else block in main.get_data_for_report? Main issue with that kind of function is that forecast data has a different structure than observation and aggregate data. Another option could be adding get_values(session, start, end) methods to the datamodel objects, but that's inconsistent with our current approach.

maybe not necessary if we make a ForecastAggregate object

> can we add a live aggregate or two that the test user can access?

I think there are two already? '458ffc27-df0b-11e9-b622-62adb5fd6af0' and 'd3d1e8e5-df1b-11e9-b622-62adb5fd6af0'

> are there issues with quality flags that should be addressed now?

dunno. The aggregate quality flag is currently just OR of individual observation flags

File: solarforecastarbiter/datamodel.py (hunk 2 above)
Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py]: probably change it to ``aggregate`` for clarity. Maybe the base class of ``ForecastObservation`` and ``ForecastAggregate`` has a property like ``data_object`` that is either an observation or aggregate for those functions that don't care about the difference.

**Comment by wholmgren** [comment]
Yeah, I found the ids in dev. I had been running the manual tests against prod which does not have any aggregates to test against. 

Latest commit adds `APISession.get_values`. I expect we'll want to write similar if/elif statements in a few places, so better to have it here. If we really don't need it in multiple places then fine to remove.

**Comment by wholmgren** [comment]
@alorenzo175 can we add a test aggregate forecast with interval_length = 60 minutes? the existing forecast is interval_length 5 minutes, but aggregate interval_length is 60 minutes, so the aggregated values cannot be compared to the forecast

**Comment by alorenzo175** [comment]
done 49220780-76ae-4b11-bef1-7a75bdc784e3. Won't survive a restart of dev mysql until my PR for jobs is in, but should be fine.

**Comment by wholmgren** [comment]
Thanks for adding the forecast.

Code now works when run with the following metadata:

```
{
    "name": "test",
    "report_parameters": {
        "start": "2019-04-01T00:00:00Z",
        "end": "2019-04-30T00:00:00Z",
        "metrics": [
            "mae",
            "mbe",
            "rmse"
        ],
        "categories": [
            "total",
            "year",
            "month",
            "date",
            "day",
            "weekday",
            "hour"
        ],
        "object_pairs": [
            {
                "forecast": "49220780-76ae-4b11-bef1-7a75bdc784e3",
                "aggregate": "458ffc27-df0b-11e9-b622-62adb5fd6af0"
            },
            {
                "forecast": "11c20780-76ae-4b11-bef1-7a75bdc784e3",
                "observation": "123e4567-e89b-12d3-a456-426655440000"
            }
      ]
    }
  }
```

Will changing the metadata specification break the ability to access existing reports?

I'll wait for a review of the new API before writing tests.

**Comment by alorenzo175** [comment]
> Will changing the metadata specification break the ability to access existing reports?

Not sure that it matters right now though. If you care, you could tackle report versioning...

File: solarforecastarbiter/datamodel.py [hunk 3]
```diff
@@ -985,7 +1005,7 @@ class ProcessedForecastObservation(BaseModel):
     parameters
     """
     # do this instead of subclass to compare objects later
-    original: ForecastObservation
+    original: Union[ForecastObservation, ForecastAggregate]
```
Comment (alorenzo175) [inline solarforecastarbiter/datamodel.py]: I think the Report forecast_observations also needs this Union

File: solarforecastarbiter/reports/figures.py [hunk 4]
```diff
@@ -56,16 +56,16 @@ def construct_fx_obs_cds(fx_values, obs_values):
 
 def _obs_name(fx_obs):
     # TODO: add code to ensure obs names are unique
-    name = fx_obs.observation.name
-    if fx_obs.forecast.name == fx_obs.observation.name:
+    name = fx_obs.data_object.name
+    if fx_obs.forecast.name == fx_obs.data_object.name:
         name += ' Observation'
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures.py]: I'm not sure where this shows up, but you don't want + 'Aggregate' for aggregates?

**Comment by lboeman** [comment]
>         "object_pairs": [
>             {
>                 "forecast": "49220780-76ae-4b11-bef1-7a75bdc784e3",
>                 "aggregate": "458ffc27-df0b-11e9-b622-62adb5fd6af0"
>             },
>             {
>                 "forecast": "11c20780-76ae-4b11-bef1-7a75bdc784e3",
>                 "observation": "123e4567-e89b-12d3-a456-426655440000"
>             }
>       ]

Do we all agree this is the structure we want here? If so, I can start working on the API to get this working on the dev server.

**Comment by wholmgren** [comment]
I'm happy with it.

**Comment by wholmgren** [comment]
@alorenzo175 ready for review but there is a failure in `test_api.py::test_apisession_post_raw_report_processed_data` that I can't figure out. Thoughts on that would be helpful.

**Comment by alorenzo175** [comment]
It was a test issue where the  order of what was returned by request mock was wrong

**Comment by wholmgren** [comment]
Thanks. I think the only remaining untested new code is some conditional renaming in `figures.py` and we can skip that.

**Pull request merged**